File: packages/common/src/events.ts
import type { Order, OrderType, Side, Symbol } from './types.js';

/**
 * Engine events. These are the SOURCE OF TRUTH for the exchange — written
 * to the Redis Stream `engine.events` in order. Every other service builds
 * its state by consuming these.
 *
 * Each event has a monotonic `seq` assigned by the engine when it is
 * appended to the log.
 */
export type EngineEvent =
  | OrderAcceptedEvent
  | OrderRejectedEvent
  | OrderCanceledEvent
  | TradeEvent;

export interface OrderAcceptedEvent {
  kind: 'OrderAccepted';
  seq: number;
  ts: number;
  order: Order;
}

export interface OrderRejectedEvent {
  kind: 'OrderRejected';
  seq: number;
  ts: number;
  clientOrderId: string;
  userId: string;
  reason:
    | 'INVALID_PRICE'
    | 'INVALID_QTY'
    | 'DUPLICATE_CLIENT_ORDER_ID'
    | 'POST_ONLY_WOULD_CROSS'
    | 'FOK_NOT_FILLABLE'
    | 'UNKNOWN_SYMBOL';
}

export interface OrderCanceledEvent {
  kind: 'OrderCanceled';
  seq: number;
  ts: number;
  orderId: string;
  userId: string;
  symbol: Symbol;
  side: Side;
  price: bigint;
  remaining: bigint;        // qty that was on the book at cancel time
  reason: 'USER' | 'IOC_REMAINDER' | 'STP' | 'MARKET_NO_LIQUIDITY';
}

export interface TradeEvent {
  kind: 'Trade';
  seq: number;
  ts: number;
  tradeId: string;
  symbol: Symbol;
  price: bigint;            // execution price (always the maker's resting price)
  qty: bigint;              // filled qty (base asset base units)
  // Aggressor is the side that TOOK liquidity (the incoming order).
  aggressor: Side;
  takerOrderId: string;
  takerUserId: string;
  takerOrderType: OrderType;
  makerOrderId: string;
  makerUserId: string;
}

/**
 * JSON-safe serialization for events: bigints become decimal strings.
 * We use this both for the Redis Stream payload and the WS wire format.
 */
export function serializeEvent(ev: EngineEvent): string {
  return JSON.stringify(ev, (_k, v) =>
    typeof v === 'bigint' ? v.toString() : v
  );
}

const BIGINT_KEYS = new Set([
  'price',
  'qty',
  'remaining',
]);

export function deserializeEvent(json: string): EngineEvent {
  return JSON.parse(json, (k, v) => {
    if (BIGINT_KEYS.has(k) && typeof v === 'string') return BigInt(v);
    return v;
  });
}

File: packages-go/common/events.go
package common

import (
	"encoding/json"
	"fmt"
)

type EngineEvent interface {
	Kind() string
	Serialize() (string, error)
}

type OrderAcceptedEvent struct {
	Seq   int64  `json:"seq"`
	TS    int64  `json:"ts"`
	Order *Order `json:"order"`
}

func (e *OrderAcceptedEvent) Kind() string { return "OrderAccepted" }

func (e *OrderAcceptedEvent) Serialize() (string, error) {
	return marshalTagged(struct {
		Kind string `json:"kind"`
		*OrderAcceptedEvent
	}{e.Kind(), e})
}

type RejectReason string

const (
	RejectReasonInvalidPrice           RejectReason = "INVALID_PRICE"
	RejectReasonInvalidQty             RejectReason = "INVALID_QTY"
	RejectReasonDuplicateClientOrderID RejectReason = "DUPLICATE_CLIENT_ORDER_ID"
	RejectReasonPostOnlyWouldCross     RejectReason = "POST_ONLY_WOULD_CROSS"
	RejectReasonFOKNotFillable         RejectReason = "FOK_NOT_FILLABLE"
	RejectReasonUnknownSymbol          RejectReason = "UNKNOWN_SYMBOL"
)

type OrderRejectedEvent struct {
	Seq           int64        `json:"seq"`
	TS            int64        `json:"ts"`
	ClientOrderID string       `json:"clientOrderId"`
	UserID        string       `json:"userId"`
	Reason        RejectReason `json:"reason"`
}

func (e *OrderRejectedEvent) Kind() string { return "OrderRejected" }

func (e *OrderRejectedEvent) Serialize() (string, error) {
	return marshalTagged(struct {
		Kind string `json:"kind"`
		*OrderRejectedEvent
	}{e.Kind(), e})
}

type CancelReason string

const (
	CancelReasonUser              CancelReason = "USER"
	CancelReasonIOCRemainder      CancelReason = "IOC_REMAINDER"
	CancelReasonSTP               CancelReason = "STP"
	CancelReasonMarketNoLiquidity CancelReason = "MARKET_NO_LIQUIDITY"
)

type OrderCanceledEvent struct {
	Seq       int64        `json:"seq"`
	TS        int64        `json:"ts"`
	OrderID   string       `json:"orderId"`
	UserID    string       `json:"userId"`
	Symbol    Symbol       `json:"symbol"`
	Side      Side         `json:"side"`
	Price     int64        `json:"price,string"`
	Remaining int64        `json:"remaining,string"`
	Reason    CancelReason `json:"reason"`
}

func (e *OrderCanceledEvent) Kind() string { return "OrderCanceled" }

func (e *OrderCanceledEvent) Serialize() (string, error) {
	return marshalTagged(struct {
		Kind string `json:"kind"`
		*OrderCanceledEvent
	}{e.Kind(), e})
}

type TradeEvent struct {
	Seq            int64     `json:"seq"`
	TS             int64     `json:"ts"`
	TradeID        string    `json:"tradeId"`
	Symbol         Symbol    `json:"symbol"`
	Price          int64     `json:"price,string"`
	Qty            int64     `json:"qty,string"`
	Aggressor      Side      `json:"aggressor"`
	TakerOrderID   string    `json:"takerOrderId"`
	TakerUserID    string    `json:"takerUserId"`
	TakerOrderType OrderType `json:"takerOrderType"`
	MakerOrderID   string    `json:"makerOrderId"`
	MakerUserID    string    `json:"makerUserId"`
}

func (e *TradeEvent) Kind() string { return "Trade" }

func (e *TradeEvent) Serialize() (string, error) {
	return marshalTagged(struct {
		Kind string `json:"kind"`
		*TradeEvent
	}{e.Kind(), e})
}

func marshalTagged(v any) (string, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func DeserializeEvent(data string) (EngineEvent, error) {
	var raw struct {
		Kind string `json:"kind"`
	}
	if err := json.Unmarshal([]byte(data), &raw); err != nil {
		return nil, err
	}

	switch raw.Kind {
	case "OrderAccepted":
		var ev OrderAcceptedEvent
		if err := json.Unmarshal([]byte(data), &ev); err != nil {
			return nil, err
		}
		return &ev, nil
	case "OrderRejected":
		var ev OrderRejectedEvent
		if err := json.Unmarshal([]byte(data), &ev); err != nil {
			return nil, err
		}
		return &ev, nil
	case "OrderCanceled":
		var ev OrderCanceledEvent
		if err := json.Unmarshal([]byte(data), &ev); err != nil {
			return nil, err
		}
		return &ev, nil
	case "Trade":
		var ev TradeEvent
		if err := json.Unmarshal([]byte(data), &ev); err != nil {
			return nil, err
		}
		return &ev, nil
	default:
		return nil, fmt.Errorf("unknown event kind: %s", raw.Kind)
	}
}
